tasm-lib/src/arithmetic/u128/lt_u128.rs

```diff
@@ -0,0 +1,231 @@
+
+                call {lt_u64}
+                // _ r3 r2 l3 l2 (l1 || l0  <  r1 || r0)
+
+                dup 4
+                dup 4
+                dup 4
+                dup 4
+                {&compare_u64}
+                // _ r3 r2 l3 l2 (l1 || l0  <  r1 || r0) (l3 || l2  ==  r3 || r2)
+
+                mul
+                // _ r3 r2 l3 l2 ((l1 || l0  <  r1 || r0) * (l3 || l2  ==  r3 || r2))
+
+                place 4
+                // _ ((l1 || l0  <  r1 || r0) * (l3 || l2  ==  r3 || r2)) r3 r2 l3 l2
+
+                call {lt_u64}
+                // _ ((l1 || l0  <  r1 || r0) * (l3 || l2  ==  r3 || r2)) (l3 || l2  <  r3 || r2)
+
+                add
+                // _ (l < r)
+                return
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+
+    use itertools::Itertools;
+    use num::bigint::ToBigUint;
+    use rand::rngs::StdRng;
+    use rand::thread_rng;
+    use rand::Rng;
+    use rand::RngCore;
+    use rand::SeedableRng;
+    use triton_vm::prelude::bfe;
+    use triton_vm::prelude::BFieldElement;
+
+    use crate::test_helpers::test_rust_equivalence_given_execution_state;
+    use crate::traits::closure::Closure;
+    use crate::traits::closure::ShadowedClosure;
+    use crate::traits::rust_shadow::RustShadow;
+    use crate::InitVmState;
+
+    use super::*;
+
+    impl LtU128 {
+        fn prepare_stack(&self, lhs: u128, rhs: u128) -> Vec<BFieldElement> {
+            let init_stack = self.init_stack_for_isolated_run();
+            let rhs_limbs = rhs
+                .to_biguint()
+                .unwrap()
+                .iter_u32_digits()
+                .pad_using(4, |_| 0)
+                .collect_vec();
+            let lhs_limbs = lhs
+                .to_biguint()
+                .unwrap()
+                .iter_u32_digits()
+                .pad_using(4, |_| 0)
+                .collect_vec();
+
+            [
+                init_stack,
+                rhs_limbs.into_iter().map(|l| bfe!(l)).rev().collect_vec(),
+                lhs_limbs.into_iter().map(|l| bfe!(l)).rev().collect_vec(),
+            ]
+            .concat()
+        }
+    }
+
+    impl Closure for LtU128 {
+        fn rust_shadow(&self, stack: &mut Vec<triton_vm::prelude::BFieldElement>) {
+            let mut left_limbs = vec![];
+            for _ in 0..4 {
+                left_limbs.push(stack.pop().unwrap().value() as u32);
+            }
+            let mut right_limbs = vec![];
+            for _ in 0..4 {
+                right_limbs.push(stack.pop().unwrap().value() as u32);
+            }
+
+            let right = right_limbs
+                .into_iter()
+                .rev()
+                .fold(0u128, |acc, limb| (acc << 32) | (limb as u128));
+            let left = left_limbs
+                .into_iter()
+                .rev()
+                .fold(0u128, |acc, limb| (acc << 32) | (limb as u128));
+
+            stack.push(bfe!((left < right) as u64));
+        }
```

Comment:
Similarly, `BFieldCodec` can be leveraged for the rust shadowing:

```rust
fn rust_shadow(&self, stack: &mut Vec<BFieldElement>) {
    fn pop_u128(stack: &mut Vec<BFieldElement>) -> u128 {
        let len = u128::static_length().unwrap();
        let limbs = (0..len).map(|_| stack.pop().unwrap()).collect_vec();
        *u128::decode(&limbs).unwrap()
    }

    let left = pop_u128(stack);
    let right = pop_u128(stack);
    stack.push(bfe!((left < right) as u64));
}
```

Or if you want to avoid repeating `u128` this often:

```rust
fn rust_shadow(&self, stack: &mut Vec<BFieldElement>) {
    fn pop_bigint<BI: BFieldCodec>(stack: &mut Vec<BFieldElement>) -> BI {
        let len = BI::static_length().unwrap();
        let limbs = (0..len).map(|_| stack.pop().unwrap()).collect_vec();
        *BI::decode(&limbs).unwrap()
    }

    let left = pop_bigint::<u128>(stack);
    let right = pop_bigint::<u128>(stack);
    stack.push(bfe!((left < right) as u64));
}
```